# Cool T-Shirts Inc

Cool T-Shirts Inc. has asked you to analyze data on visits to their website. Your job is to build a funnel, which is a description of how many people continue to the next step of a multi-step process.

In this case, our funnel is going to describe the following process:

A user visits CoolTShirts.com
A user adds a t-shirt to their cart
A user clicks “checkout”
A user actually purchases a t-shirt

In [1]:
import pandas as pd

In [3]:
visits = pd.read_csv('visits.csv',
                     parse_dates=[1])
cart = pd.read_csv('cart.csv',
                   parse_dates=[1])
checkout = pd.read_csv('checkout.csv',
                       parse_dates=[1])
purchase = pd.read_csv('purchase.csv',
                       parse_dates=[1])

In [5]:
print(visits.head())
print(cart.head())
print(checkout.head())
print(purchase.head())

                                user_id          visit_time
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00
                                user_id           cart_time
0  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00
1  4397f73f-1da3-4ab3-91af-762792e25973 2017-05-27 01:35:00
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663 2017-03-04 10:38:00
3  b594862a-36c5-47d5-b818-6e9512b939b3 2017-09-27 08:22:00
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974 2017-07-26 15:48:00
                                user_id       checkout_time
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d 2017-06-25 09:29:00
1  4ac186f0-9954-4fea-8a27-c081e428e34e 2017-04-07 20:11:00
2  3c9c78a7-124a-4b77-8d2e-e1926e011e7d 2017-07-13 11:38:00
3  89fe330a-8966-4756-8f7c-3bdbcd47279a 

In [11]:
# Combining visits and cart
visits_cart = pd.merge(visits,cart,how='left')
print(len(visits_cart))
print(len(visits_cart[visits_cart.cart_time.isnull()]))
visits_cart.head()

2000
1652


user_id          visit_time  \
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00   
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00   
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00   

            cart_time  
0                 NaT  
1 2017-01-26 14:44:00  
2 2017-08-20 08:31:00  
3                 NaT  
4                 NaT

In [13]:
# percent of users who visited Cool T-Shirts Inc. ended up not placing a t-shirt in their cart
print(f'Percent of users who visited Cool T-Shirts Inc. ended up not placing a t-shirt in their cart is {float((1652/2000)*100)}.')

Percent of users who visited Cool T-Shirts Inc. ended up not placing a t-shirt in their cart is 82.6


In [17]:
# Combining cart and checkout
cart_checkout = pd.merge(cart,checkout,how='left')
print(len(cart_checkout))
print(len(cart_checkout[cart_checkout.checkout_time.isnull()]))
cart_checkout.head()

482
122


user_id           cart_time  \
0  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00   
1  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00   
2  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00   
3  4397f73f-1da3-4ab3-91af-762792e25973 2017-05-27 01:35:00   
4  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663 2017-03-04 10:38:00   

        checkout_time  
0 2017-11-07 21:14:00  
1 2017-11-07 20:50:00  
2 2017-11-07 21:11:00  
3                 NaT  
4 2017-03-04 11:04:00

In [18]:
# percentage of users put items in their cart, but did not proceed to checkout
print(f'Percentage of users put items in their cart, but did not proceed to checkout {float((122/482)*100)}.')

Percentage of users put items in their cart, but did not proceed to checkout 25.311203319502074.


In [22]:
# Combining ckeckout and purchase
checkout_purchase = pd.merge(checkout,purchase,how='left')
print(len(checkout_purchase))
print(len(checkout_purchase[checkout_purchase.purchase_time.isnull()]))
checkout_purchase.head()

598
101


user_id       checkout_time  \
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d 2017-06-25 09:29:00   
1  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d 2017-06-25 09:29:00   
2  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d 2017-06-25 09:29:00   
3  4ac186f0-9954-4fea-8a27-c081e428e34e 2017-04-07 20:11:00   
4  3c9c78a7-124a-4b77-8d2e-e1926e011e7d 2017-07-13 11:38:00   

        purchase_time  
0 2017-06-25 09:49:00  
1 2017-06-25 09:56:00  
2 2017-06-25 09:57:00  
3                 NaT  
4                 NaT

In [23]:
#  percentage of users proceeded to checkout, but did not purchase a t-shirt
print(f'percentage of users proceeded to checkout, but did not purchase a t-shirt is {float((101/598)*100)}')

percentage of users proceeded to checkout, but did not purchase a t-shirt is 16.88963210702341


# Weakest Funnel
Percent of users who visited Cool T-Shirts Inc. ended up not placing a t-shirt in their cart is 82.6
Percentage of users put items in their cart, but did not proceed to checkout 25.311203319502074.
Percentage of users proceeded to checkout, but did not purchase a t-shirt is 16.88963210702341.

According to our analysis weakest funnel is Visits to Cart has the highest percentage of users not completing it .
So, Cool T-Shirts Inc Should work on their add to cart feature and work on it.


In [29]:
# Average Time to Purchase
all_data = visits.merge(cart).merge(checkout).merge(purchase)
all_data['Time_to_purchase'] = all_data.purchase_time  - all_data.visit_time
average_time = all_data['Time_to_purchase'].mean()
print(f'Average Time for a customer to make Purchase is {average_time} ')
all_data.head()

Average Time for a customer to make Purchase is 0 days 00:43:53.360160965 


user_id          visit_time  \
0  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
1  486480e2-98c3-4d51-8f4b-b1c07228ce84 2017-01-27 16:34:00   
2  3ccdaf69-2d30-40de-b083-51372881aedd 2017-01-08 20:21:00   
3  3ccdaf69-2d30-40de-b083-51372881aedd 2017-01-08 20:21:00   
4  3ccdaf69-2d30-40de-b083-51372881aedd 2017-01-08 20:21:00   

            cart_time       checkout_time       purchase_time Time_to_purchase  
0 2017-01-26 14:44:00 2017-01-26 14:54:00 2017-01-26 15:08:00  0 days 00:44:00  
1 2017-01-27 16:44:00 2017-01-27 17:10:00 2017-01-27 17:12:00  0 days 00:38:00  
2 2017-01-08 20:38:00 2017-01-08 20:52:00 2017-01-08 21:02:00  0 days 00:41:00  
3 2017-01-08 20:38:00 2017-01-08 20:52:00 2017-01-08 21:21:00  0 days 01:00:00  
4 2017-01-08 20:38:00 2017-01-08 20:52:00 2017-01-08 21:02:00  0 days 00:41:00

In [30]:
# Average Time for a customer to make Purchase is 0 days 00:43:53.360160965